File: frontend/src/types/workflow.ts
export type WorkflowStatus = 'pending' | 'running' | 'completed' | 'failed';

export type AgentRole =
  | 'planner'
  | 'researcher'
  | 'technical_architect'
  | 'developer'
  | 'reviewer'
  | 'final_answer';

export type WorkflowValue = string | Record<string, unknown> | null;

export interface HealthResponse {
  status: string;
  version: string;
  environment: string;
}

export interface WorkflowStep {
  id: string;
  role: AgentRole;
  name: string;
  description?: string | null;
  input?: WorkflowValue;
  output?: WorkflowValue;
  status: WorkflowStatus;
  error?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  duration_ms?: number | null;
}

export interface WorkflowRun {
  id: string;
  input?: WorkflowValue;
  task?: string;
  output?: WorkflowValue;
  status: WorkflowStatus;
  steps: WorkflowStep[];
  final_answer?: string | null;
  error?: string | null;
  created_at?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  duration_ms?: number | null;
  total_duration_ms?: number | null;
}

export interface RunWorkflowRequest {
  task: string;
}

File: backend/app/models/workflow.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import uuid4

from pydantic import BaseModel, Field

from app.models.enums import AgentRole, WorkflowStatus


class WorkflowStep(BaseModel):
    """One agent step inside an orchestration workflow."""

    id: str = Field(default_factory=lambda: str(uuid4()))
    role: AgentRole
    name: str
    description: str | None = None
    input: str | dict[str, Any] | None = None
    output: str | dict[str, Any] | None = None
    status: WorkflowStatus = WorkflowStatus.PENDING
    error: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = Field(default=None, ge=0)


class WorkflowRun(BaseModel):
    """Runtime representation of a full multi-agent workflow execution."""

    id: str = Field(default_factory=lambda: str(uuid4()))
    input: str | dict[str, Any]
    status: WorkflowStatus = WorkflowStatus.PENDING
    steps: list[WorkflowStep] = Field(default_factory=list)
    error: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = Field(default=None, ge=0)


class WorkflowResult(BaseModel):
    """Final result returned after a workflow finishes."""

    id: str = Field(default_factory=lambda: str(uuid4()))
    input: str | dict[str, Any] | None = None
    output: str | dict[str, Any] | None = None
    status: WorkflowStatus
    steps: list[WorkflowStep] = Field(default_factory=list)
    error: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = Field(default=None, ge=0)


__all__ = [
    "WorkflowStatus",
    "WorkflowStep",
    "WorkflowRun",
    "WorkflowResult",
]
